# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

In [2]:
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)

In [3]:
type(df)

pandas.core.frame.DataFrame

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [8]:
df.head(4) 

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [5]:
df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [9]:
df[615:620:]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [10]:
df.dtypes

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [14]:
with pd.option_context("display.precision", 2):
    display(df['gross income'].describe(percentiles=[0.01,0.25, 0.5, 0.75, 0.99]))

count    1000.00
mean       15.38
std        11.71
min         0.51
1%          0.91
25%         5.92
50%        12.09
75%        22.45
99%        45.25
max        49.65
Name: gross income, dtype: float64

## Відповідь: 
75 персентиль дарівнює 22,45.
### Давайте розберемось з данними , що нам повернутись від коду:
* так як колонака в нас називається валовий дохід, але значення від 0,51 до 49,65 , тому припускаю, що це відсоткове значення. 
* всього маємо 1000 рядків даних, середнє значення 15,38 , тобто середня прибутковість 15,38%
* стандартне відхилення 11,71, не дуже сильно розкинуті данні відносно середнього, тобто +-11.71 коливається наша прибутковість рахунків
### Tепер персентилі :
* рахунків які приносять менше 5,92% прибутків, меньше чверті, половина рахунків приносять прибуток від 5,92% до 22,45%, ще чверть рахунків високоприбуткові(в розрізі цього набору данних) і їх прибутковість вище 22,45% і сягає до 49,65%.
*  також помічено, що максимум і мінімум не дуже відрісняються від 99 і 1 персентиля відповідно, тому можу примустити, що викидів тут немає.
### Висновок:
  75 персентиль показує нам, що три чверті данних(більшість) містять інформацію про рахунки прибутковість яких менше 22,45%. 75% інвойсів приносять меньшн або 22,45% валового прибутку.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [21]:
df.City.value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [20]:
df.City.value_counts(normalize=True)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [41]:
df[df.City == 'Yangon']['gross income'].mean()

np.float64(14.874001470588235)

In [42]:
df[df.City == 'Mandalay']['gross income'].mean()

np.float64(15.232024096385542)

In [43]:
df[df.City.isin(['Yangon', 'Mandalay'])].groupby('City')['gross income'].mean()

City
Mandalay    15.232024
Yangon      14.874001
Name: gross income, dtype: float64

### Зважаючи на інформацію яку ми вже знаємо в цих двох місцях не нуже відчутна різниця інвойсів Янгон 340, Мандалай 332, проте ми бачимо що там де більше інвойсів чомусь відсоток прибутку менший, чим це спричинено?
#### Гіпотеза 1
В одному місті можливо куплять товари які більше прибуткові у чеку. наприклад є категорії товарів з більшою маржею або щось таке.
#### Гіпотеза 2
В одному інвойсі можлива покупка декількох товарів, тому така різниця ,кількість інвойсім менше, а прибутку більше, бо в 1 чеку 15 товарів.
#### Гіпотеза 3
Можливо в одному місці проживають бульше клієнтів які мають статус мембер і вони або мають знижку , або навпаки мають накопичувальну картку і куплять частіше саме там .

### Висновок
Схиляють до теорії про кількість продуктів у чеку, проглянувки ще раз фрйм помітила що вал.прибуток(вп) = такс5%, тотал = ціна за штуку + такс5% * на кількість товару , і ВП пропорційно до тоталу більшується або зменшується, отже на тотал впливає кількість товару і на ВП впливає кількість товару.
В Янгоні просто продали меншу кількість товару , а в Мандалай продали більшу кількість товару (по декілька в одному чеку)

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [49]:
g = [df['gross margin percentage'].unique()]
if len(g) == 1:
    print('У всіх всіх інвойсах однакова маржа')
else:
    print( 'Маржа відрізняється для різних інвойсів')

У всіх всіх інвойсах однакова маржа


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [52]:
df[df.City == 'Yangon']['Product line'].value_counts(normalize=True)

Product line
Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: proportion, dtype: float64

In [53]:
df[df.City == 'Mandalay']['Product line'].value_counts(normalize=True)

Product line
Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: proportion, dtype: float64

#### Перевіряю свою гіпотезу по кількості товарів

In [57]:
df[df.City == 'Yangon']['Quantity'].sum()

np.int64(1859)

In [58]:
df[df.City == 'Mandalay']['Quantity'].sum()

np.int64(1820)

In [68]:
df.groupby('Product line')['Unit price'].mean().sort_values(ascending=False)

Product line
Fashion accessories       57.153652
Sports and travel         56.993253
Food and beverages        56.008851
Home and lifestyle        55.316937
Health and beauty         54.854474
Electronic accessories    53.551588
Name: Unit price, dtype: float64

In [66]:
df[df['City'] == 'Yangon'].groupby('Product line')['Quantity'].sum().sort_values(ascending=False)

Product line
Home and lifestyle        371
Sports and travel         333
Electronic accessories    322
Food and beverages        313
Fashion accessories       263
Health and beauty         257
Name: Quantity, dtype: int64

In [69]:
df[df['City'] == 'Mandalay'].groupby('Product line')['Quantity'].sum().sort_values(ascending=False)

Product line
Sports and travel         322
Health and beauty         320
Electronic accessories    316
Fashion accessories       297
Home and lifestyle        295
Food and beverages        270
Name: Quantity, dtype: int64

In [70]:
df[df['City'] == 'Mandalay'].groupby('Product line')['cogs'].sum().sort_values(ascending=False)

Product line
Sports and travel         19036.38
Health and beauty         19029.20
Home and lifestyle        16713.49
Electronic accessories    16239.47
Fashion accessories       15631.73
Food and beverages        14490.37
Name: cogs, dtype: float64

In [71]:
df[df['City'] == 'Yangon'].groupby('Product line')['cogs'].sum().sort_values(ascending=False)

Product line
Home and lifestyle        21349.71
Sports and travel         18450.19
Electronic accessories    17444.87
Food and beverages        16345.81
Fashion accessories       15554.77
Health and beauty         11997.86
Name: cogs, dtype: float64

Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)